Add a onepass dfa matcher.

ethanpailes opened the pull request:
This PR adds the onepass DFA that I've been working on. I have marked it as a WIP because there is still one failing test case that I need to dig my teeth into.

Currently the only way to active the onepass matcher is to call .onepass() on ExecBuilder. I was hoping to get this merged in without having it enabled by default at first, then I can do a seperate PR which adds more testing/benchmarking and actually enables it. That approach also provides more time to discuss how exactly the onepass matcher will get used by the execution planner. My guess is that the onepass matcher should always be used when it can be, but it may turn out that shoveling capture groups around is pricier than the DFA startup cost. 

We talked about test coverage over in #68, so I did some simple checks and found that about 300/700 of the tests in the test suite trigger the onepass matcher.

Closes #68

BurntSushi commented:
:tada: :tada: :tada: :tada: :tada:

I can't wait to find time to dig into this. It still might be a bit though.

> We talked about test coverage over in #68, so I did some simple checks and found that about 300/700 of the tests in the test suite trigger the onepass matcher.

Wow. That's awesome news. One thing that I think is important here is to create an additional test suite that _disables_ the one-pass matcher so that those 300 tests hit the other engines as well.

(How the various engines are tested desperately needs to be improved, but that's well outside the scope here I think. It's definitely slated for a rethink as I refactor regex internals.)

ethanpailes commented:
I'm probably going to get bit for posting before CI finishes, but that last commit should turn all the tests green. The issue turned out to be a silly config bug.

ethanpailes commented:
Another possible optimization: don't record `last_match` until we leave a match state. Search for "another inner loop!" in the `src/dfa.rs` source code.

ethanpailes commented:
Given that there have been a number of new features and a new minimum Rust version since I first opened this PR, I've rebased onto the current `master`. I've also squashed the various polishing patches that I added, so there are now only two patches in the PR. This is still a pretty big diff, so I understand that it will probably take time to digest (if it would help reduce the work at all I would be happy to talk through the design online or in person).

bors commented:
:umbrella: The latest upstream changes (presumably 2918aab48110dc9aa960b23901ffb7f226ac4241) made this pull request unmergeable. Please resolve the merge conflicts.

RReverser commented on regex-syntax/src/hir/interval.rs:
```diff
@@ -309,6 +309,11 @@ impl<I: Interval> IntervalSet<I> {
         true
     }
+
+    /// Returns true iff this class is empty.
```
Typo: `if`.

BurntSushi replied:
I don't think it's a typo. `iff` is parlance for "if and only if."

RReverser commented on regex-syntax/src/hir/mod.rs:
```diff
@@ -797,6 +797,16 @@ impl ClassUnicode {
         self.set.symmetric_difference(&other.set);
     }
+
+    /// Returns true iff this character class contains no characters.
```
Same typo.

RReverser commented on regex-syntax/src/hir/mod.rs:
```diff
@@ -998,6 +1008,16 @@ impl ClassBytes {
         self.set.intervals().last().map_or(true, |r| r.end <= 0x7F)
     }
+
+    /// Returns true iff this character class contains no characters.
```
Same (probably copy-paste).

BurntSushi commented on src/analysis.rs:
```diff
@@ -0,0 +1,547 @@
+                self.next()
+            }
+            _ => Some(&es[idx]),
+        }
```
@ethanpailes I _believe_ this routine uses a stack size proportional to the depth of the HIR. It seems like it should be possible to flatten this with a loop.

ethanpailes replied:
Nice catch. Thanks! I'll fix this with a loop as you suggest.

Note to future self: The degenerate case is when a bunch of concats are nested, with the child concat always coming first.

BurntSushi commented on src/analysis.rs:
```diff
@@ -0,0 +1,547 @@
+    // there is byte-level nondeterminism in the \d character
+    // class, and we care about things in the byte space rather
+    // than the character space. If you do a onepass engine at
+    // the character level, Cox's example is indeed onepass.
```
Could we add a test for said regex?

IIRC, RE2 uses a byte level representation as well. I wonder if the real reason here is that `\d` is Unicode aware by default here, but is only ASCII aware in RE2. That is, would `(?-u)(\d+)-(\d+)` be one pass?

I think overall I would expect to see quite a bit more tests that check the top level `is_onepass` routine here. :-)

BurntSushi replied:
Yeah, the RE2 onepass matcher is definitely byte oriented because RE2 doesn't have any character level byte codes. Even in this crate, the pike VM can match Unicode character classes without character level byte codes. For RE2, see: https://github.com/google/re2/blob/e6acac839be126c8a20a20183f20a23fa9f4895e/re2/onepass.cc#L459-L524

The Go onepass matcher---along with all of the matchers in Go---are however character oriented. There is no byte oriented matching in Go's regexp engine IIRC.

ethanpailes commented:
The work on this branch started off with me under the impression that I was just engaged in porting the onepass matcher from RE2 to this crate, but as I learned more along the way the algorithm I settled on drifted further away from the one used there. In order to avoid confusion it might be worth giving it a new name, such as "unambigious_dfa". "onepass" sounds cooler, but it is probably worth being descriptive.

BurntSushi commented:
@ethanpailes SGTM. How much work would it be to write up a description of how the detection works in a comment in the code? That would help me a lot.

I don't think it needs to be done in this PR, but one thing worth thinking about is whether the algorithm can be built incrementally as HIR are constructed. This would avoid the need for having a second pass over the HIR. You can see how I'm doing this with a variety of other attributes by looking at the various HIR constructors. e.g., https://github.com/rust-lang/regex/blob/cd6f1e2d6dcea8a76c18045e336f5e360e7ad784/regex-syntax/src/hir/mod.rs#L361 --- This is something I'd be happy to adapt the current code to, so I'm in no way asking for you to rewrite it here, but getting your thoughts on the feasibility of the idea would be very useful!

RReverser commented:
Not sure how feasible this would be, but could one pass matches be also used for regexps which normally are not eligible due to parallel threads with capture groups, but are called with `.is_match`, `.find` or similar method which ignores capture groups anyway? I imagine this would increase memory consumption, but might be worth it in terms of speed?

BurntSushi commented:
> As for determining is_onepass all in one compiler pass when going from Ast -> Hir

Hmm, I think you might have misunderstood me. My suggestion isn't really about the translation from AST to HIR, but rather, the _construction_ of _any_ HIR. There is no "pass" here; the idea is that the property is defined inductively on the structure of the HIR. That is, given any two HIRs `x` and `y` and a known value for whether each is onepass or not, can you compute the following in time proportional to the size of its non-recursive sub-expressions?

```
Concat(x, y) => onepass(x) && onepass(y) && no_intersection(tail(x), head(y))
Alternation(x, y) => ...
Repeat(x) => ...
...
```

and so on. My guess is that there might be additional state required but that it otherwise seems plausible.

> Are you just worried about the performance/memory implications of letting the crate sprout a middle end? A separate pass seems more maintainable to me.

Computing properties upon construction of the HIR is what feels like the most ideal solution is, with respect to performance, memory and maintenance. :-) It is explicitly inductive and impossible to forget, because it's baked directly into the construction of the HIR.

I am totally fine starting with a separate pass. I don't think having a separate pass is _bad_. I can give more thoughts on the feasibility of the explicitly inductive approach once I understand the detection algorithm better.

> but we might also be able to just determine onepassness by examining the bytecode

Yeah, that is plausible. For now, I like the idea of using the HIR. I don't have a good feel for what role the bytecode will play in analysis passes. In the past, I've found the bytecode more difficult to do analysis on, but it's been a while. My intent is to rethink the bytecode completely.

BurntSushi commented on src/analysis.rs:
```diff
@@ -0,0 +1,547 @@
+            // TODO(ethan):yakshaving factor the determination of when
+            //     a regex accepts_empty out into a separate function,
+            //     so that we don't compute the whole first set when we
+            //     don't need to.
```
Have you seen? https://docs.rs/regex-syntax/0.6.2/regex_syntax/hir/struct.Hir.html#method.is_match_empty

(That's an example of a property that is computed inductively as a HIR is constructed.)

ethanpailes commented on src/analysis.rs:
```diff
@@ -0,0 +1,547 @@
+        &HirKind::Alternation(ref es) => {
+            let mut fset = FirstSet::empty();
+            for e in es {
+                fset.union(&fset_of(e));
```
Yeah, this is pretty unforgivable on my part. The solution is definitely going to have to include caching results somewhere, so I've started work rewriting this module as an extra thing that gets constructed while building Hirs like you've suggested elsewhere. I should have it fixed soon.

BurntSushi replied:
@ethanpailes Awesome! I just want to re-iterate that I am OK with doing structural recursion here for the time being. Literal anaylsis already does this, so it's not introducing a new regression.

The problem that could potentially arise with doing this at HIR construction time is that I think you'll need to keep a copy of fsets for every HIR expression, and that could prove to be quite precarious. This isn't something I realized at the outset when I suggested it. So it's not clear to me that it is _definitely_ the path we want to take.

Sorry for the back and forth on this, but I'm learning as I go. :)

ethanpailes replied:
I was kinda worried about the additional memory usage, but the fact that FirstSets are represented with a ClassBytes instead of a [u8; 256] does help a bit. What's the use case to worry about for an expanded Hir memory footprint? I would think that there would be more memory used during execution of the regex (at least in the worst case), and I don't think adding a ClassBytes to each Hir node will take up more memory than the 3 Programs and a OnePass that make up an ExecReadOnly. Does the Hir make it into a Regex object somewhere that I havn't noticed?

If I was trigger happy in doing the re-write I can revert that last patch and just try to either add some more ephemeral caching or ditch the visitor to get rid of the quadratic behavior.

BurntSushi replied:
It's certainly hand wavy, sure, but for example, a naive implementation would add an fset to every character in a pattern. Moreover, the size of an fset is proportion to the size of the fsets of its sub-expressions, so the growth there could be worrisome.

But honestly, I don't know. It's just a concern. This is definitely something I'm going to have to play with.

Looking at your latest commit, I don't think the implementation is actually what we want anyway. If you're looking at sub-sub-expressions _during construction_ (such as with `NestedConcat`), then that's basically wrong and defeats the purpose of computing this property at construction.

Perhaps the best path forward here is to just do the straight-forward thing and use structural recursion. No visitors. No construction time properties. I can sort this out later, perhaps as part of sorting out literal analysis.

ethanpailes replied:
I think we are saved from storing an fset per character by the Literal hir node, and the size of an fset shouldn't grow any bigger a bit less than 256 bytes. Plus larger fsets should have a smaller memory foot print, unless I'm off the mark on how ClassBytes works.

The NestedConcat issue is a bit trickier. I added NestedConcat while I was working to get the whole test suite to pass. It is only needed for some pretty weird regex, so I can go back to revisit the idea to see if there is a way to get rid of it. (Also isn't strong induction still induction?)

BurntSushi replied:
> I think we are saved from storing an fset per character by the Literal hir node, and the size of an fset shouldn't grow any bigger a bit less than 256 bytes. Plus larger fsets should have a smaller memory foot print, unless I'm off the mark on how ClassBytes works.

I'm afraid this isn't true. Look at the HIR for a simple expression:

```
$ regex-debug hir 'abc'
Hir {
    kind: Concat(
        [
            Hir {
                kind: Literal(
                    Unicode(
                        'a'
                    )
                ),
                info: HirInfo {
                    bools: 1
                }
            },
            Hir {
                kind: Literal(
                    Unicode(
                        'b'
                    )
                ),
                info: HirInfo {
                    bools: 1
                }
            },
            Hir {
                kind: Literal(
                    Unicode(
                        'c'
                    )
                ),
                info: HirInfo {
                    bools: 1
                }
            }
        ]
    ),
    info: HirInfo {
        bools: 1
    }
}
```

Notice that _every character_ has its own `HirInfo` value. [Your commit adds an fset to `HirInfo`](https://github.com/rust-lang/regex/pull/467/commits/6355ce519b115f1b4c3bcbaac1b824e1e5151309#diff-18cb88398f4c23fb21519633d53f4aefR1658), which means every character will have an fset, which means we have quite a bit of overheard per character. Not only is this problematic in terms of size, but this will actually wind up being a separate independent allocation per character.

Now, this isn't necessarily a death blow to computing the onepass property on construction, but I think it's pretty clear that it requires more work than I thought---potentially even refactoring the HIR itself.

> (Also isn't strong induction still induction?)

Only logically. If you implement induction by actually using every previous step, then you've effectively lost one of the benefits of building this stuff by construction because you still wind up doing work proportional to the depth of the HIR, and that in turn means you've implementing a quadratic algorithm. The really key idea here is that the construction of an HIR expression should _never_ take time proportional to the depth of its sub-expressions. It should always take time proportional to the number of direct sub-expressions.

I'd like to nudge you towards moving back to structural recursion where things are a bit simpler to reason about. I'd also like to say that I don't think we can really permit quadratic algorithms _anywhere_ in this code, unless we explicitly bound them by some small (heuristically chosen) constant.

ethanpailes replied:
I didn't notice that Literals in Hir only have one character now (I was still thinking of them as the old Expr::Literal). The allocation-per-character thing is an issue. We could solve it by special casing an fset that contains only one byte, but that seems a bit hacky.

I'm going to look into avoiding the use of NestedConcat to fix the worst-case quadratic behavior. I wasn't saying that NestedConcat doesn't make the runtime worst-case quadratic, just quibbling unnecessarily over the word inductive.

BurntSushi replied:
I don't think you need to fix the use of nested concat because that isn't enough to make this approach work. Could you please go back to the structural recursion implementation as an explicit analysis pass?

ethanpailes commented on src/analysis.rs:
```diff
@@ -0,0 +1,547 @@
+
+/// Check if a list of first sets is incompatable.
+///
+/// O(n^2), but n will usually be quite small.
```
I'll fix it by maintaining a union of fsets seen so far in a loop and checking against that.

ethanpailes commented:
@RReverser I'm not sure I understand the optimization you are talking about; maybe you can expand on it a bit. The onepass dfa has full support for capture groups. I think if you call .is_match or .find on a regex with capture groups the normal lazy dfa will get used, but its been a little while since I read the execution planner.

RReverser commented on src/analysis.rs:
```diff
@@ -0,0 +1,547 @@
+    }
+
+    fn check_alternation(&mut self, es: &[Hir]) {
+        self.0 = self.0 && !fsets_clash(&es.iter().collect::<Vec<_>>());
```
Not sure if it would be possible to change `fsets_clash` to always accept `&[Hir]` or that would require cloning in another callsite, but at least you can make it generic over both with something like:

```rust
fn fsets_clash<H: Borrow<Hir>>(es: &[H]) -> bool
```

ethanpailes replied:
That sounds like exactly the sort of type magic I was trying to find. I just pushed a patch with my macro hack, but I'll give that a try soon.

ethanpailes replied:
It looks like Borrow is more about smart pointers than what I want to do here.

RReverser replied:
No, not necessarily. In my experience it's most common usecase is exactly accepting owned vs references to some type.

ethanpailes commented on src/onepass.rs:
```diff
@@ -0,0 +1,1187 @@
+        // instructions which get states.
+        let mut state_edge = vec![0];
+        while let Some(i) = state_edge.pop() {
+            state_edge.extend(self.inst_trans(i, &mut forwards)?);
```
Extending `state_edge` with a vector here will cause an allocation per instruction. I should just pass a mutable reference to `state_edge` into `inst_trans` and mutate it directly instead.

ethanpailes commented on src/onepass.rs:
```diff
@@ -0,0 +1,1187 @@
+
+        // Iterate over the children, visiting lower priority
+        // children first.
+        let mut resume = match &self.prog[inst_idx] {
```
If the cost of allocating this resume queue for each Bytes instruction is a problem, this might be a candidate for a SmallVec. Probably not worth the dependency without some measurements.

ethanpailes commented on src/onepass.rs:
```diff
@@ -0,0 +1,1187 @@
+    /// The arrow flows in a funny direction because we want the jobs
+    /// with no dependencies to live at the roots of the DAG so that
+    /// we can process them first.
+    fn perform_forward(&mut self, fwd: Forward) {
```
I didn't see this whole forwarding scheme in the RE2 code, and I think that the reason boils down to the fact that there is DFA states in the RE2 encoding occupy a fixed amount of space in the transition table, so you can point to a state that has not yet been computed pretty easily. To keep the DFA state encoding small, RE2 devotes more of its equivalent of StatePointers to flags and control. One result of this is that RE2's onepass engine only supports a finite number of capture groups.

It might be worth trying to improve the OnePassCompiler performance by simplifying our DFA encoding a bit, but I think that would probably fit better in a separate PR.

ethanpailes replied:
Another concern with forwarding is the time complexity. The graph algorithms involved are O(N*E) where N is the number of instructions and E is the edges between those instructions. It shouldn't be bad if the number of edges is O(N), which feels like it ought to be true, but I don't have an off the cuff proof about why. Its at least worth thinking about.

RReverser commented:
@ethanpailes Right, I suppose I misread description from the original issue

> Detecting whether a regex is deterministic or not. A regex is said to be deterministic if, at any point during the execution of a regex program, at most one thread can lead to a match state

as talking about captures but now I see that it talks about matching in general.

ethanpailes commented:
I just reverted the commit checking for onepassness at Hir construction time and replaced it with a version of the analysis.rs module that uses direct recursion on the Hir. It still uses strong induction instead of weak induction because of the NestedConcat iterator, but it shouldn't hurt the time complexity because I don't recurse on the nested concatenation Hir nodes, only their children.

BurntSushi commented:
@ethanpailes Great, thank you!

ethanpailes commented:
@BurntSushi, I know reviewing this change is asking you to do a lot of work, and even though you mentioned trying to get the review done before launching into a big crate-level refactoring I don't want this PR to block you from working on your ideas. I'm perfectly happy to rebase on top of major changes and re-work this PR. The diff for this PR looks rather invasive because of the integration with the execution planner and whatnot, but most of the work went into `onepass.rs`, which is fairly self-contained. Instruction format changes would be the only thing that really required big changes to `onepass.rs`, and even then it should be fairly manageable unless the semantics change a lot. 

I'm also just selfishly excited to see how you end up re-working the compiler, because I'm sure that I will learn a lot.

ethanpailes commented:
Ah, ok. I was worried that you already had a grand vision brewing and this change was getting in the way.

I tried to make the code a clean as possible, but it is just inherently a big ball of low-level state. If you make an attempt at reviewing it and run into trouble because of the volume I would be happy to respond to questions, refactor to make the code clearer, and add comments to explain the guts. Another thing is that if the style/structure of the code is a barrier to understanding I would be happy to re-work it (for example maybe my decision to use the iterator protocol to view a DAG in topological order has a more direct, clearer solution that I didn't think of). I trust your judgement about what makes code grokkable significantly more than my own. I'm happy to respond to partial reviews of that sort for however long it takes for you to either get comfortable with the code or decide that it doesn't fit in with the regex crate, and again I don't want this PR blocking other changes.

bors commented:
:umbrella: The latest upstream changes (presumably 8f9ca96a844b07e6b378573ec6fbac6cf9fe5430) made this pull request unmergeable. Please resolve the merge conflicts.

ethanpailes commented:
@BurntSushi I'm going to close the PR for now, though if you ever want to re-open it later, I'd be happy to work to get it merged.

The code in this PR is very algorithmically dense, and while there are some real performance wins they are only for a subset of cases. The longer a regex gets, the less likely it is to be deterministic. Also, because of the lazy DFA, the onepass matcher would probably only bring big wins when looking for capture groups. On the other hand, re2 does include a onepass matcher, and Russ Cox is pretty clever.

Given that this PR has been open for years and you've attempted to give it a proper review several times without being able to really digest it all, I think it is mostly just a drain on maintainer resources and you are already stretched pretty thin between all your different crates. (Also closing it now would give the PR queue for regex inbox zero for the first time in years :slightly_smiling_face: ).

BurntSushi commented:
@ethanpailes Thanks! It's probably fine to close this for now, you're right. Basically, I'm still working on #656, and my plan there is still to incorporate a one-pass matcher inside of `regex-automata`. What I'd like to do is think about the problem from first principles for myself, then review this PR and see what I can pull out and move into `regex-automata`.

Optimizing the capture group case is rather important and comes up more often than you might think. One common example is parsing out the sections of each line in a log file. Right now, that's a pretty common case that is very slow with the `regex` crate that I think something like the one-pass matcher could help out with a lot.

Believe me, I've had this on my mind all this time, but you're right, I've consistently bounced off of this. But I'm still hoping to work my way toward it. :-)

(regex-automata is coming along fairly nice. I have almost the entire DFA component done with tons of improvements to the Thompson NFA compiler. That will add another type of matching engine to the regex crate. It will also provide a lower level "expert" API for thornier cases such as stream searching or more powerful RegexSet-like functionality. Namely, when using a full DFA, you can search for multiple regexes simultaneously and get their start/end offsets, just like how the aho-corasick crate works.)

BurntSushi commented:
@ethanpailes So I've started to let my mind wonder about how to do a one-pass regex engine in regex-automata and finally circled back around to your PR to check out what you did in more detail (particularly with respect to the DFA).

My suspicion is that the technique of using a DFA for this in the context of the regex crate is probably dead-on-arrival unfortunately. I'm sorry if I led you down that path. While I haven't taken any measurements, just looking at the compilation phase, for example, suggests that it's probably going to spend a lot of time _and_ memory building the DFA. The thing with DFAs is that we don't just want to avoid them because of the possibility of exponential state blow-up, but also because they are pigs in general. There's not much you can do about the fact that a DFA state is just plain huge (even with the byte class optimization) without sacrificing some portion of its speed advantages. This is why the lazy DFA has its cache capped at a couple MB (IIRC). We really just do not want to be using that much memory per regex.

While I haven't measured anything here---so I'm mostly just guessing---it kinda looks like the one-pass compiler here is probably way too expensive. To a first approximation, the NFA compiler itself is already almost too expensive, particularly when it comes to dealing with huge Unicode character classes.

I haven't looked at the analysis phase yet, so it's still possible to salvage that component I think.

ethanpailes commented:
@BurntSushi that makes sense.

OnePass regex tend to be on the smaller side since the more is in a regular expression the more chances of an ambiguous transition. It's been a few years since I wrote this, so I don't have everything swapped in, but I think that despite the complexity here there is a chance the space usage won't be too bad because of the lack of exponential blow up and general small size of regex that manage to qualify as OnePass.

Adding a space limit to the compiler (or maybe trying to do it in the analysis pass) might be a way to salvage this work, but I understand if you don't want to open that can of worms.

If you're just using this as an artefact to study for what you do for real, I think it is worth mentioning that this implementation gets a lot of its complexity from trying to handle arbitrary numbers of capture expressions. To do this, states with control codes take up double the space of normal states, which makes the math for figuring out where a given state starts a little harder than if they were all fixed size. The re2 implementation just packs all its control flags in the high order bits of the address words and does not suffer from this problem and much of the complexity it implies.

BurntSushi commented:
Aye, thanks for the response. The main problem here is somewhat fundamental. A single state in a DFA takes `size_of::<idtype>() * alphabet.len()` bytes all on its own, and this adds up pretty quickly unfortunately.

Keep in mind that compilation time is also an issue.

The other thing I'll likely try to address is whether there's a way to make things like Unicode-aware `\w` fit into the one-pass model. Not sure if it's possible, but if Unicode character classes kill the one-pass optimization, then it makes it a lot less useful unfortunately.

ethanpailes commented:
> The main problem here is somewhat fundamental. A single state in a DFA takes size_of::<idtype>() * alphabet.len() bytes all on its own, and this adds up pretty quickly unfortunately.

Yeah, I can't think of a way around that except the byte class compression that this already does. I guess maybe you could make it lazy? That seems like a lot of effort for onepass regex though.

What about ditching the idea of a DFA entirely and just writing an NFA VM that just takes advantage of the fact that there can be no non-determinism in one pass regex? I think such an NFA engine would not have to deal with juggling thread state arrays like in the Pike VM or any sort of backtracking. There might be some perf wins to be had there, though it might be that the backtracking VM is good enough since it will just never backtrack and already skips the cost of the thread state arrays.
